Fetch Kafka Connect, MM2, Bridge trusted certificates Secret only once for each reference

venkatesh2090 opened the pull request:
### Type of change

_Select the type of your PR_

Enhancement / new feature

### Description

Fixes #11972

Updated when Secrets for trusted certificates are fetched and used in KafkaConnect, KafkaMirrorMaker2, and KafkaBridge assembly operators. It was fetched twice - once when the combined CA cert secret is generated, and again when the hash is calculated with certs and auth.

Refactored the reconcile loop such that it is only fetched once per reference.

**KafkaConnect**
It is fetched in `tlsTrustedCertsSecret` and returned with the method, which is passed into the `generateAuthHash` method.

**KafkaBridge**
A similar pattern is used here - but since a new secret isn't generated, it is fetched directly in the loop and passed to the hash method.

**KafkaMirrorMaker2**
Since there are multiple certs for each mirror's target and source, all the certs are fetched and stored in a `HashMap`. It relies on the fact that each cluster in MM2 has a unique alias. It is then passed to both secret generation and auth hash methods when needed.

I have also removed 2 tests in `ReconcilerUtilsTest` which checked if the `authTlsHash` method fetched the secrets correctly. That method doesn't fetch trusted cert secrets anymore. It expects certs to be passed in as parameter. 

Disclosure: The operator tests were generated using AI and modified. I have reviewed it thoroughly before committing.

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [x] Write tests
- [x] Make sure all tests pass
- [x] Update documentation
- [x] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md
- [ ] Supply screenshots for visual changes, such as Grafana dashboards

venkatesh2090 committed:
Remove redundant calls to fetch tls secrets.

Refactored tlsTrustedCertsSecret to return a list of secrets instead of the concatenated version.

Signed-off-by: Venkatesh Kannan <venkatesprasad512@gmail.com>

venkatesh2090 committed:
Add tests.

Also fixed auth hash in MM2.

Signed-off-by: Venkatesh Kannan <venkatesprasad512@gmail.com>

venkatesh2090 committed:
refactor: move auth hash method back to where it was

Signed-off-by: Venkatesh Kannan <venkatesprasad512@gmail.com>

venkatesh2090 committed:
fix stylcheck error

Signed-off-by: Venkatesh Kannan <venkatesprasad512@gmail.com>

venkatesh2090 committed:
Fix NullPointerExceptions

Signed-off-by: Venkatesh Kannan <venkatesprasad512@gmail.com>

venkatesh2090 committed:
refactor indentation

Signed-off-by: Venkatesh Kannan <venkatesprasad512@gmail.com>

venkatesh2090 committed:
Add coverage for oauth tls secret

Signed-off-by: Venkatesh Kannan <venkatesprasad512@gmail.com>

scholzj commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/ReconcilerUtils.java:
```diff
@@ -531,7 +530,7 @@ public static Future<String> trustedCertificates(Reconciliation reconciliation,
                             return Future.succeededFuture();
                         } else {
-                            return Future.succeededFuture(String.join("\n", certificates.list()));
+                            return Future.succeededFuture(certificates.list());
```
Does the list have a fixed order? That is important to make sure the hash does not change randomly.

tinaselenge commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaMirrorMaker2AssemblyOperator.java:
```diff
@@ -128,7 +131,15 @@ protected Future<KafkaMirrorMaker2Status> createOrUpdate(Reconciliation reconcil
                 .compose(i -> serviceOperations.reconcile(reconciliation, namespace, mirrorMaker2Cluster.getServiceName(), mirrorMaker2Cluster.generateService()))
                 .compose(i -> serviceOperations.reconcile(reconciliation, namespace, mirrorMaker2Cluster.getComponentName(), mirrorMaker2Cluster.generateHeadlessService()))
-                .compose(i -> tlsTrustedCertsSecret(reconciliation, namespace, mirrorMaker2Cluster))
+                .compose(i -> updateMM2ClusterCertificateMap(reconciliation, mirrorMaker2Cluster, clusterCerts))
```
I'm not sure if I understand these changes? Can you please provide some explanation?

venkatesh2090 replied:
I think I implemented the auth hash method incorrectly. I meant to use the map like

```java
clusterCert.get(cluster.getAlias())
```

and then pass it to `ReconcilerUtils.authTlsHash`. I meant to replicate this behaviour https://github.com/strimzi/strimzi-kafka-operator/pull/12241/files/63d40adaf074ca0d50be8e7b4bc294a58f8b3d73#diff-c6d9b371d64b172f4e174d61f4a6b81ed129010c41a016f0bbf29dddc62df009L204-L207

```java
                        .map(cluster -> {
                            List<CertSecretSource> trustedCertificates = cluster.getTls() == null ? List.of() : cluster.getTls().getTrustedCertificates();
                            return ReconcilerUtils.authTlsHash(secretOperations, namespace, cluster.getAuthentication(), trustedCertificates);
                        }).collect(Collectors.toList())
                )
```

by doing this

```diff
@@ -210,12 +211,14 @@ public class KafkaMirrorMaker2AssemblyOperator extends AbstractConnectOperator<K
     private Future<Integer> generateAuthHash(String namespace, KafkaMirrorMaker2Cluster mirrorMaker2Cluster, Map<String, List<String>> clusterCert) {
         Promise<Integer> authHash = Promise.promise();
 
-        List<String> certificates = clusterCert.values().stream().flatMap(List::stream).toList();
         Future.join(mirrorMaker2Cluster
                         .clusters()
                         .stream()
-                        .map(KafkaMirrorMaker2ClusterSpec::getAuthentication)
-                        .map(auth -> ReconcilerUtils.authTlsHash(secretOperations, namespace, auth, certificates))
+                        .map(cluster -> {
+                            KafkaClientAuthentication auth = cluster.getAuthentication();
+                            List<String> certificates = clusterCert.get(cluster.getAlias());
+                            return ReconcilerUtils.authTlsHash(secretOperations, namespace, auth, certificates);
+                        })
                         .collect(Collectors.toList())
                 )
```


> but we previously just used target's clusters' certs to generate the hash

I'm not sure about this part. Didn't it use all the clusters - source and target, to generate the auth hash via the `KafkaMirrorMaker2Cluster.clusters()` method? Although I didn't implement that part correctly.

tinaselenge commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/ReconcilerUtils.java:
```diff
@@ -513,15 +512,15 @@ public static Future<Integer> authTlsHash(SecretOperator secretOperations, Strin
      * @return  Certificates extracted from the Secrets
      */
-    public static Future<String> trustedCertificates(Reconciliation reconciliation, SecretOperator secretOperations, List<CertSecretSource> certificateSources)   {
+    public static Future<List<String>> trustedCertificates(Reconciliation reconciliation, SecretOperator secretOperations, List<CertSecretSource> certificateSources)   {
```
Does this really need to be changed to a List? Can the authTlsHash not be computed from the final String instead of a List?

venkatesh2090 replied:
Cool. I will refactor to make this simpler. The only place this list of certs is used is for the auth hash calculation.

tinaselenge reviewed (commented):
Thanks for the PR. I left some comments. I think we need to leave KafkaBridge related changes out of this PR as mentioned below.

venkatesh2090 committed:
Update javadoc, and remove unused code.

Signed-off-by: Venkatesh Kannan <venkatesprasad512@gmail.com>

venkatesh2090 committed:
Fix tests after rebasing main

Signed-off-by: Venkatesh Kannan <venkatesprasad512@gmail.com>

venkatesh2090 committed:
Refactored trustedCertificates method to return a String of concatenated certificates.

This simplifies authTlsHash implementation by using hashCode on the certBundle instead of adding all hashes.

Signed-off-by: Venkatesh Kannan <venkatesprasad512@gmail.com>

venkatesh2090 committed:
Missed a blank string check

Signed-off-by: Venkatesh Kannan <venkatesprasad512@gmail.com>

venkatesh2090 commented:
@tinaselenge @ppatierno I have done some refactoring and rebased against latest main. Would appreciate another review

tinaselenge commented:
Thanks @venkatesh2090 . Just to let you know that there are changes that will probably conflict with your PR now:
https://github.com/strimzi/strimzi-kafka-operator/pull/12735 
https://github.com/strimzi/strimzi-kafka-operator/pull/12728

Once both of these are merged, you might want to rebase and see if these changes would simplify your changes as truststore certificate handling will become more consistent across Bridge, Connect and MM2.

katheris commented:
Discussed on community call on 14.5.2026: @venkatesh2090 apologies for us failing to get back to this. As you have likely seen there's a bunch of changes being made in this area by myself and @tinaselenge. If you are still up for continuing on this task then let us know and Tina can help you to rebase your changes to make sense with the changes she's making. If you don't have time for this task anymore no worries, just let us know and Tina has agreed to take your PR forwards and get it merged.

Sorry again for us dropping the ball here.

venkatesh2090 commented:
No problem. I haven't had the time to look into this recently either. I'll try to look over this weekend. If I don't get back in a week, feel free to take over.

venkatesh2090 committed:
Fix after rebasing

Signed-off-by: Venkatesh Kannan <venkatesprasad512@gmail.com>

venkatesh2090 committed:
Use existing tlsTrustedCertsSecret method merged from main

Signed-off-by: Venkatesh Kannan <venkatesprasad512@gmail.com>

venkatesh2090 committed:
Remove unused variable and imports

Signed-off-by: Venkatesh Kannan <venkatesprasad512@gmail.com>

venkatesh2090 committed:
Update javadoc for tlsTrustedCertsSecret

Signed-off-by: Venkatesh Kannan <venkatesprasad512@gmail.com>

venkatesh2090 commented:
@tinaselenge I have rebased onto main with your changes now. I didn't have many conflicts. The only one I had was in Bridge Assembly Operator, where I reused one of the newer methods.

Would appreciate another review.

katheris commented:
The unit test failed in a UserOperator test so I have restarted the job to see if it is an intermittent failure

katheris commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/ReconcilerUtils.java:
```diff
@@ -488,7 +486,7 @@ public static Future<Integer> authTlsHash(SecretOperator secretOperations, Strin
     }
 
     /**
-     * Gets trusted certificates from Secrets and merges them into a single String.
```
What's the reason for this change?

tinaselenge replied:
The update to this comment looks like a mistake as the return type is still String?

katheris replied:
@venkatesh2090 can you revert this change please?

katheris commented on cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaBridgeAssemblyOperatorTest.java:
```diff
@@ -771,4 +775,82 @@ public void testUpdatedBridgeConfigurationHash(VertxTestContext context) {
+    @Test
+    public void testTlsSecretsFetchedOnlyOncePerReference(VertxTestContext context) {
+        String kbName = "foo";
+        String kbNamespace = "test";
```
You should use the static variables at the top of this class: `NAME` and `NAMESPACE`

katheris commented on cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaBridgeAssemblyOperatorTest.java:
```diff
@@ -771,4 +775,82 @@ public void testUpdatedBridgeConfigurationHash(VertxTestContext context) {
+                        )
+                    .endTls()
+                .endSpec()
+                .build();
```
Rather than creating a lot of this from scratch you can re-use the basic bridge already initialised:

```suggestion
        KafkaBridge bridgeWithTls = new KafkaBridgeBuilder(BRIDGE)
                .editSpec()
                    .withNewTls()
                        .withTrustedCertificates(
                                new CertSecretSourceBuilder()
                                        .withSecretName("shared-tls-secret")
                                        .withCertificate("ca.crt")
                                        .build(),
                                new CertSecretSourceBuilder()
                                        .withSecretName("shared-tls-secret")
                                        .withCertificate("ca2.crt")
                                        .build()
                        )
                    .endTls()
                .endSpec()
                .build();
```

katheris commented on cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaBridgeAssemblyOperatorTest.java:
```diff
@@ -771,4 +775,82 @@ public void testUpdatedBridgeConfigurationHash(VertxTestContext context) {
+        when(mockBridgeOps.updateStatusAsync(any(), any())).thenReturn(Future.succeededFuture());
+
+        when(mockDcOps.scaleDown(any(), eq(kbNamespace), any(), anyInt(), anyLong())).thenReturn(Future.succeededFuture());
+        when(mockDcOps.scaleUp(any(), eq(kbNamespace), any(), anyInt(), anyLong())).thenReturn(Future.succeededFuture());
```
Are these two mocks needed? Does the test actually scale anything?

venkatesh2090 replied:
Seems like they both are being used.

https://github.com/strimzi/strimzi-kafka-operator/blob/e26ca546bf555b4b2c3d4775aa88f08829717f3e/cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaBridgeAssemblyOperator.java#L109

https://github.com/strimzi/strimzi-kafka-operator/blob/e26ca546bf555b4b2c3d4775aa88f08829717f3e/cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaBridgeAssemblyOperator.java#L124

katheris commented on cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaBridgeAssemblyOperatorTest.java:
```diff
@@ -771,4 +775,82 @@ public void testUpdatedBridgeConfigurationHash(VertxTestContext context) {
+        Checkpoint async = context.checkpoint();
+        ops.reconcile(reconciliation)
+                .onComplete(context.succeeding(v -> context.verify(() -> {
+                    verify(mockSecretOps, times(2)).getAsync(eq(kbNamespace), eq("shared-tls-secret"));
```
I wonder if it would be better to have two different secrets listed? Then we can ensure we are calling each reference once

tinaselenge replied:
I wonder if we should make this test a bit more rigorous. We could add 2 different secrets, one with 2 keys, and another one with just one key. And then check if one secret is accessed twice and another secret is accessed once only. This will make sure, we are retrieving all certificates in the same secret and when there is just one certificate in a secret, we access that only once.

katheris commented on cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectAssemblyOperatorPodSetTest.java:
```diff
@@ -2128,4 +2129,99 @@ public void testManualRollingUpdateWithSuppressedFailure(VertxTestContext contex
+        Checkpoint async = context.checkpoint();
+        ops.reconcile(new Reconciliation("test-trigger", KafkaConnect.RESOURCE_KIND, NAMESPACE, NAME))
+                .onComplete(context.succeeding(v -> context.verify(() -> {
+                    verify(mockSecretOps, times(2)).getAsync(eq(NAMESPACE), eq("shared-tls-secret"));
```
Similarly to the bridge test, I'm curious why both are referencing the same secret rather than different secrets?

venkatesh2090 replied:
I can't remember now if I'm being honest. This was probably generated by LLMs. I'll change this in all other places as well.

katheris commented on cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/ReconcilerUtilsTest.java:
```diff
@@ -457,75 +428,13 @@ void getHashFailure() {
     }
 
-    @Test
-    void getHashForPattern(VertxTestContext context) {
```
I think these tests are still valid, although the `authTlsHash` method is no longer fetching the certificates, this test is actually testing the logic inside the `getTrustedCertificateAsync` method around how we combine the fetched Secrets into a single String. So I think we should update these tests to call `getTrustedCertificateAsync` directly instead and rename them but keep them

venkatesh2090 replied:
Seems like `getTrustedCertificateAsync` is private. I'm not sure what the sentiment around making a private method public / package-private for the purposes of testing is. If it is fine, I'd be happy to make it visible to the tests.

katheris replied:
You can make it package private and mark it with a `/* test */` comment

venkatesh2090 replied:
replacing `getHashForPattern` test still feels redundant if I'm being honest. That functionality is also tested in `testTrustedCertificates` via `ReconcilerUtils.trustedCertificates`.

https://github.com/strimzi/strimzi-kafka-operator/blob/d366a7d5cf50085b36c23d86cbb57a9d43c7122f/cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/ReconcilerUtilsTest.java#L590-L593

I can replace `getHashPatternNotMatching` by just using `ReconcilerUtils.trustedCertificates` and keeping `getTrustedCertificateAsync` private, since that part isn't tested anywhere.

I feel like `trustedCertificates` tests are also testing for most  `getTrustedCertificateAsync` functionality anyways.

Would appreciate your thoughts.

venkatesh2090 replied:
@katheris let me know what your opinions are when you get a chance.

katheris replied:
Hi @venkatesh2090, sorry I missed the response to this. I'm happy with you updating the tests to call `trustedCertificates` directly rather than changing the scope of the methods just for tests

venkatesh2090 replied:
Thanks @katheris. Updated

katheris commented on cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaMirrorMaker2AssemblyOperatorPodSetTest.java:
```diff
@@ -1333,4 +1338,107 @@ private KafkaConnectApi createConnectClientMock() {
+        Checkpoint async = context.checkpoint();
+        ops.reconcile(new Reconciliation("test-trigger", KafkaMirrorMaker2.RESOURCE_KIND, NAMESPACE, NAME))
+                .onComplete(context.succeeding(v -> context.verify(() -> {
+                    verify(mockSecretOps, times(2)).getAsync(eq(NAMESPACE), eq("shared-tls-secret"));
```
Similar comment here to elsewhere about referencing the same secret twice.

venkatesh2090 replied:
I have made some changes and added some comments here. Want to bring attention to these changes as well.

Seems like tls secret for the "target" cluster in MM2 is being retrieved twice because getTls() on MM2 returns the "target" cluster's tls configuration. And it is once again retrieved when iterating through all the clusters in the same method.

https://github.com/venkatesh2090/strimzi-kafka-operator/blob/298cdde01e44e491c3886964474b7730907e3e51/cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaMirrorMaker2AssemblyOperator.java#L198-L214

katheris replied:
Ah I see @tinaselenge is that an oversight in your changes? What's the reason for having the target TLS config listed as both `ca.crt` and `target-alias-ca.crt`?

tinaselenge replied:
It was intentionally done that way because we first configure the underlying Connect cluster with the certificate keyed by generic `ca.crt` which is independent from MM2 configuration (https://github.com/strimzi/strimzi-kafka-operator/blob/fdbd17194242a201da279979967a33bf11adfea2/cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaConnectConfigurationBuilder.java#L98). Then we configure Connectors separately, for that we need a way to differentiate the certs for each cluster, so that source and target are configured with the correct certs (https://github.com/strimzi/strimzi-kafka-operator/blob/fdbd17194242a201da279979967a33bf11adfea2/cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaMirrorMaker2Connectors.java#L260). For this, it was simpler to use the alias as the key for the certs. So this results in copying the target certs into the internal secret twice but with the different keys.

katheris replied:
Ok thanks @tinaselenge, makes sense I'm happy with us leaving the test here as it is then

katheris commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaMirrorMaker2AssemblyOperator.java:
```diff
@@ -295,6 +309,20 @@ private Future<Void> maybeUpdateMirrorMaker2Status(Reconciliation reconciliation
+            return ReconcilerUtils.trustedCertificates(reconciliation, secretOperations, trustedCerts)
+                .compose(certs -> {
+                    if (certs != null) {
+                        // Duplicate check is done in KafkaMirrorMaker2Cluster::validateAndUpdateToNewAPI
```
This is now referencing a method that no longer exists, I've tried to look and I can't find that duplication checking is being done elsewhere instead, so I think I need some more time to investigate and understand what the behaviour is now if there are duplicate aliases.

tinaselenge commented:
@venkatesh2090 thank you so much for rebasing the PR for the Bridge related changes. I'm quite happy with the changes around Connect and Bridge now. However, I still think MirrorMaker part is a bit complicated since it is handled differently from Bridge and Connect. 

The second PR I mentioned #12735 is not merged yet, which is why you don't see conflicts. I'm wondering if #12735 would simplify the changes around MirrorMaker and make it more consistent with the changes you have done with Connect and Bridge. 

For example, if you look at the new tlsTrustedCertsSecret() method [here](https://github.com/tinaselenge/strimzi-kafka-operator/blob/64d3913c1075ef343cf1f77ae1cf640bf8d855de/cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaMirrorMaker2AssemblyOperator.java#L195), it will create a map for the certificates. Then you could make this method return the map as a result, and then move the generateAuthHash() call below the tlsTrustedCertsSecret call in the reconciliation chain just like what you did for Connect and Bridge. 

If this PR gets merged first before #12735, we will have to refactor things around again afterwards. So I think we could coordinate our PRs to avoid the need to refactor.

katheris reviewed (commented):
Hi @venkatesh2090, thanks for rebasing this. I've taken a look at the changes, I'm also happy with the code changes for Connect and Bridge, however I need more time to investigate the MirrorMaker changes since some of the code that you were relying on has been removed in 1.0.0.

I've added some comments around the tests, so please take a look at those.

In terms of the MirrorMaker changes, given Tina's comment about her PR and the fact that the alias duplication checking code is now gone I think the best approach here is to remove the MM2 logic from this PR and open a separate PR just for MM2. That way we can get the Connect and Bridge changes in and review the MM2 changes in a more focused PR.

@venkatesh2090 @tinaselenge what do you think?

tinaselenge commented:
> In terms of the MirrorMaker changes, given Tina's comment about her PR and the fact that the alias duplication checking code is now gone I think the best approach here is to remove the MM2 logic from this PR and open a separate PR just for MM2. That way we can get the Connect and Bridge changes in and review the MM2 changes in a more focused PR.

I think that's a good idea to separate MM2 into another PR. Then this PR will not be blocked by MM2 related discussions/changes.

venkatesh2090 commented:
@tinaselenge @katheris I am happy to separate MM2 into another PR. I don't mind waiting for #12735 to be merged and rebasing my MM2 changes against it in the future.

venkatesh2090 committed:
Addressing comments: reusing existing variables for simplification

Signed-off-by: Venkatesh Kannan <venkatesprasad512@gmail.com>

venkatesh2090 committed:
Addressing comments: Improve tests

Signed-off-by: Venkatesh Kannan <venkatesprasad512@gmail.com>

ppatierno commented:
@venkatesh2090 could you please resolve conflicts before we have another pass to review this? Thanks!

venkatesh2090 committed:
Simplify after rebasing

Signed-off-by: Venkatesh Kannan <venkatesprasad512@gmail.com>

venkatesh2090 commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaMirrorMaker2AssemblyOperator.java:
```diff
@@ -191,8 +192,7 @@ protected Future<Boolean> delete(Reconciliation reconciliation) {
      */
-    @Override
-    protected Future<Void> tlsTrustedCertsSecret(Reconciliation reconciliation, String namespace, KafkaConnectCluster mirrorMaker2Cluster) {
+    protected Future<Void> tlsTrustedCertsSecret(Reconciliation reconciliation, String namespace, KafkaConnectCluster mirrorMaker2Cluster, Map<String, String> clusterCerts) {
```
Want to bring attention to changes in this method. It is a bit janky. I changed a method that @tinaselenge made. I copy the map of mm2 certs into a variable and reused it in auth hash.

I couldn't think of a clever way to incorporate the new changes. Also this modifies the method signature, hence it is no longer overriding the AbstractConnectOperator method

tinaselenge replied:
I understand the issue. I wonder if this method can return all the certificates from all the clusters as a single string that way it matches the method it is trying to override. Would computing hash from each cluster's certificates, or computing hash from all certificates from all clusters be different? 

So for example, this method could return `.map(String.join("\n", secretData.values()));`

and then generateAuthHash can look like:
```
    private Future<Integer> generateAuthHash(String namespace, KafkaMirrorMaker2Cluster mirrorMaker2Cluster, String certs) {
        Promise<Integer> authHash = Promise.promise();

        // Single string containing all cluster certs)
        int certsHash = certs == null ? 0 : certs.hashCode();

        // Get authentication hash for each cluster and combine them
        Future.join(mirrorMaker2Cluster
                        .clusters()
                        .stream()
                        .map(cluster -> ReconcilerUtils.authTlsHash(secretOperations, namespace, cluster.getAuthentication(), List.of()))
                        .collect(Collectors.toList())
                )
                .onSuccess(hashes -> {
                    int authsHash = hashes.<Integer>list()
                        .stream()
                        .mapToInt(i -> i)
                        .sum();
                    authHash.complete(certsHash + authsHash);
                }).onFailure(authHash::fail);

        return authHash.future();
    }
```

This needs to be of course tested to make sure the behaviour is still the same and pods get rolled when certificates are updated etc.

venkatesh2090 replied:
This makes sense. The hashes would be different if we do it like this after all. I'll get back when I need help with testing this. Thanks.

venkatesh2090 replied:
Actually seems like I won't need to combine the two hashes outside `authTlsHash`.

Each of the auth type branch is already adding the cert's hash to the auth's hash. e.g.

https://github.com/strimzi/strimzi-kafka-operator/blob/f52d46302d1f8ec7eed0a640d1f1596da4f12903/cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/ReconcilerUtils.java#L472

So I can just pass in the common certs hash to `authTlsHash` and get a different hash each time instead of `List.of()`.

```diff
-                        .map(cluster -> ReconcilerUtils.authTlsHash(secretOperations, namespace, cluster.getAuthentication(), List.of()))
+                        .map(cluster -> ReconcilerUtils.authTlsHash(secretOperations, namespace, cluster.getAuthentication(), certs))
```

To be tested still.

venkatesh2090 replied:
Should I add a test to `KafkaMirrorMaker2AssemblyOperatorPodSetTest` to test if pods are rolled out when the auth hash annotation changes? Trying to figure out if there is a more suitable class I can test this in. I couldn't find one that tests this explicitly

venkatesh2090 replied:
Added to `KafkaMirrorMaker2AssemblyOperatorPodSetTest`

tinaselenge replied:
` .map(cluster -> ReconcilerUtils.authTlsHash(secretOperations, namespace, cluster.getAuthentication(), certs))
`
Would this mean, the final hash will contain hash of combined certs (passed to the method as argument) multiple times? 

`certs` that is passed as an argument contains certs for both clusters, then you iterate through each cluster to get the hash for the authentications. But if you pass `certs` to `ReconcilerUtils.authTlsHash()` method for each cluster, you would end up with something like 
`sourceAuthHash + (combined source and target certs hash) + targetAuthHash +  (combined source and target certs hash)`. 
The changes I proposed was passing an empty list of certs to `ReconcilerUtils.authTlsHash()`, you would get `sourceAuthHash + targetAuthHash + (combined source and target certs hash)`.

Maybe it doesn't matter too much as long as we detect that the hash has changed, so this change could be simpler even though adding the certs hash twice to final hash. @katheris do you have thoughts on this?

venkatesh2090 replied:
That's right. I was thinking it is less change and nothing more than that. I thought the auth hash would change with either fucntion. But I don't know enough about hash function to justify my changes. I am happy to change it to your suggestion or @katheris's if hashes are more likely to collide this way.

tinaselenge replied:
Actually I think code change wise, it's not too different so let's not add certs hash twice. 
We can call  ReconcilerUtils.authTlsHash with empty list for `certs` argument, to calculate the authentication hash for each cluster and then add the `certs` hash in the end. Basically, what I have suggested above.

tinaselenge commented on cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaMirrorMaker2AssemblyOperatorPodSetTest.java:
```diff
@@ -1344,4 +1349,126 @@ private KafkaConnectApi createConnectClientMock() {
+
+    @Test
+    public void testTlsSecretsFetchedOnlyOncePerReference(VertxTestContext context) {
+        // Create MM2 with TLS configured for both source and target clusters using the same secret
```
They are actually using different secrets? target is using shared-tls-secret and source is using shared-tls-secret-2?

venkatesh2090 replied:
Right. Didn't pay attention to the comment. Will update

venkatesh2090 replied:
Or maybe I'll just remove the comment. It seems redundant

tinaselenge reviewed (commented):
Thank you for the update. I've left a few more comments.

venkatesh2090 committed:
Simplified hashing functionality

Signed-off-by: Venkatesh Kannan <venkatesprasad512@gmail.com>

katheris commented on cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaBridgeAssemblyOperatorTest.java:
```diff
@@ -771,4 +775,83 @@ public void testUpdatedBridgeConfigurationHash(VertxTestContext context) {
+                .build();
+
+        when(mockBridgeOps.get(eq(NAMESPACE), eq(NAME))).thenReturn(bridgeWithTls);
+        when(mockBridgeOps.getAsync(eq(NAMESPACE), eq(NAME))).thenReturn(Future.succeededFuture(bridgeWithTls));
```
I don't think you should need to mock both get and getAsync, I think only getAsync should be being called?

katheris reviewed (commented):
Thanks @venkatesh2090 this is looking good now, just one small nit on a test and I added a comment that I want Tina to look at, but I think we're very close now to this being ready to go in

venkatesh2090 committed:
Remove unnecessary mock and fix comment

Signed-off-by: Venkatesh Kannan <venkatesprasad512@gmail.com>

venkatesh2090 committed:
Add test for verifying pod rollouts when secret is added or modified.

Signed-off-by: Venkatesh Kannan <venkatesprasad512@gmail.com>

venkatesh2090 committed:
Make the test more comprehensive:

Verify when secrets are new.
Then, when secrets haven't changed.
Then, when secrets were updated.

Signed-off-by: Venkatesh Kannan <venkatesprasad512@gmail.com>

tinaselenge commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaMirrorMaker2AssemblyOperator.java:
```diff
@@ -126,15 +125,15 @@ protected Future<KafkaMirrorMaker2Status> createOrUpdate(Reconciliation reconcil
                 .compose(i -> connectPodDisruptionBudget(reconciliation, namespace, mirrorMaker2Cluster))
-                .compose(i -> generateAuthHash(namespace, mirrorMaker2Cluster))
+                .compose(i -> tlsTrustedCertsSecret(reconciliation, namespace, mirrorMaker2Cluster))
+                .compose(certs -> generateAuthHash(namespace, mirrorMaker2Cluster, certs))
```
why generateAuthHash is not moving up instead of moving down tlsTrustedCertsSecret just like the other reconcilers?

venkatesh2090 replied:
No particular reason. From what I can tell, it wouldn't make much of a difference I suppose. If you want consistency, I can update all the reconcilers to be so?

tinaselenge replied:
I think you can update this reconciler to match other ones. tlsTrustedCertsSecret is generally called before the generateMetricsAndLoggingConfigMap, so let's move this generateAuthHash before generateMetricsAndLoggingConfigMap for consistency.

tinaselenge commented on cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaMirrorMaker2AssemblyOperatorPodSetTest.java:
```diff
@@ -1344,4 +1392,209 @@ private KafkaConnectApi createConnectClientMock() {
+                    List<Pod> newPods = PodSetUtils.podSetToPods(newPodSet);
+                    for (int i = 0; i < mm2Cluster.getReplicas(); i++) {
+                        assertThat(newPods.get(i), not(oldPods.get(i)));
+                        assertThat(newPods.get(i).getMetadata().getAnnotations(), not(oldPods.get(i).getMetadata().getAnnotations()));
```
can we check the annotations value if that matches exactly what we expect?

venkatesh2090 replied:
I'm not entirely sure how the revision and the revision-hash labels are calculated. That's why I only checked if the annotations map has changed from the old pods.

tinaselenge reviewed (commented):
Thanks @venkatesh2090 for addressing the comments. I just have a few more comments but I think we are nearly there.

venkatesh2090 committed:
Addressed more comments

Signed-off-by: Venkatesh Kannan <venkatesprasad512@gmail.com>

tinaselenge commented:
/gha run pipeline=regression

tinaselenge commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaBridgeAssemblyOperator.java:
```diff
@@ -112,15 +108,15 @@ protected Future<KafkaBridgeStatus> createOrUpdate(Reconciliation reconciliation
             })
             .compose(i -> isPodDisruptionBudgetGeneration ? podDisruptionBudgetOperator.reconcile(reconciliation, namespace, bridge.getComponentName(), bridge.generatePodDisruptionBudget()) : Future.succeededFuture())
-            .compose(i -> ReconcilerUtils.authTlsHash(secretOperations, namespace, auth, trustedCertificates))
+            .compose(i -> tlsTrustedCertsSecret(reconciliation, namespace, bridge))
```
nit: I guess authTlsHash should move up rather than move down for consistency?

venkatesh2090 replied:
done. must've missed this

venkatesh2090 replied:
Looks like a test fails just by moving it 😟. didn't have a chance to figure why yet. will look later

venkatesh2090 replied:
Fixed

tinaselenge reviewed (approved):
Thank you for the PR and I appreciate addressing of the comments. I left one minor comment, otherwise looks good to me :) 

The PR needs rebasing again, hopefully they are straightforward conflicts to resolve. Once that's done, I can kick off the regression tests again.

venkatesh2090 committed:
Fix tests after rebase

Signed-off-by: Venkatesh Kannan <venkatesprasad512@gmail.com>

venkatesh2090 committed:
Address comments

Signed-off-by: Venkatesh Kannan <venkatesprasad512@gmail.com>

tinaselenge commented:
/gha run pipeline=regression

katheris commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/AbstractConnectOperator.java:
```diff
@@ -275,23 +275,34 @@ protected Future<ReconcileResult<NetworkPolicy>> connectNetworkPolicy(Reconcilia
+     * Reconciles the secret that combines secrets and certificates
+     * provided for Kafka Connect truststore if TLS is enabled. Uses
+     * {@code certificates} instead of fetching secrets in {@code connect}.
+     *
```
Can you add this missing `@param` lines to the Javadoc please

katheris commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaMirrorMaker2AssemblyOperator.java:
```diff
@@ -254,23 +253,23 @@ protected Future<Void> tlsTrustedCertsSecret(Reconciliation reconciliation, Stri
-                            List<CertSecretSource> trustedCertificates = cluster.getTls() == null ? List.of() : cluster.getTls().getTrustedCertificates();
-                            return ReconcilerUtils.authTlsHash(secretOperations, namespace, cluster.getAuthentication(), trustedCertificates);
-                        }).collect(Collectors.toList())
+                        .map(cluster -> ReconcilerUtils.authTlsHash(secretOperations, namespace, cluster.getAuthentication(), ""))
```
Won't this always return 0?

katheris reviewed (commented):
I added one small nit for Javadoc, but otherwise for MM2 I'm not sure I follow the changes you've made for calculating the hash, you seem to be returning 0, I guess to keep the behaviour you had before, but I'm not sure why even make the call to the method

katheris commented:
@venkatesh2090 I've had a chat to @tinaselenge offline and realised I had misread the changes for MM2, so I understand better now the intention. However, we have been discussing further about the calculations of the hashes and really think we should be maintaining the previous behaviour where we calculate the hash for each certificate and then sum them, rather than concatenating the certs and then calculating the hash. Otherwise if the order of certificates changes then the hash will be different.

So can we request two changes please, can you update the code to pass `List<String>` to the `ReconcilerUtils.authTlsHash` method (we know you had this before - apologies!), also can you take another look at the MM2 method `tlsTrustedCerts` and update it to only return the List of certs for the clusters? I think you should be able to create a dedicated List for the return, rather than constructing it from the Secret data.

Let me or Tina know if any of this is unclear, or if you wanted we would also be happy to arrange a call to talk to you about it. We really appreciate your time on this and are still committed to helping you get this change in.

venkatesh2090 commented:
Thanks both for your commitment and offering to setup a call. Really appreciate it! While it would definitely help I am in the same timezone as you guys and can't do it during work hours unfortunately. I'll ask any clarifying questions here.

> can you update the code to pass `List<String>` to the `ReconcilerUtils.authTlsHash` method

If I do this I would have to change all other reconcilers. Unless you meant the MM2 operator `generateAuthHash` method? 

> can you take another look at the MM2 method `tlsTrustedCerts` and update it to only return the List of certs for the clusters? I think you should be able to create a dedicated List for the return, rather than constructing it from the Secret data.

I would need to return a `Map<String, String>` of cluster alias -> cert to be able to get the right cert based on what `cluster.authentication()` is used in `generateAuthHash` like I had before 

https://github.com/venkatesh2090/strimzi-kafka-operator/blob/bd41a31ca3494363729c6f16ecb51b3dc1d18e93/cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaMirrorMaker2AssemblyOperator.java#L267

---
What I have in my head right now is returning `Map<String, String>` from MM2 `tlsTrustedCertsSecret`. This map will then be used in `generateAuthHash` here

https://github.com/venkatesh2090/strimzi-kafka-operator/blob/12d7731103debc60b1f31361c173b0f0d7c4ed81/cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaMirrorMaker2AssemblyOperator.java#L130.

So instead of `certs` it would be a `certsMap` of sorts. 

Later in the `generateAuthHash` method, we get the correct cert and authentication based on the cluster alias in the Stream.

tinaselenge commented:
> If I do this I would have to change all other reconcilers. Unless you meant the MM2 operator generateAuthHash method?

I think the ordering issue is relevant to all of the reconcilers, for example if Bridge is configured with multiple trusted certificates and ordering might change when joining them into a single String, which would trigger a rolling due to hash being different even though the certificates are exactly the same. I think you originally proposed to pass `List<String>` for the certs to generateAuthHash for all the reconcilers but I overlooked this issue, so I apologise for pointing you to the wrong direction! 

> What I have in my head right now is returning Map<String, String> from MM2 tlsTrustedCertsSecret. This map will then be used in generateAuthHash here

In MirrorMaker case, you could keep a similar logic you have now. generateAuthHash does not actually need to know which certs belong to which clusters so no need of Map here. If tlsTrustedCertsSecret returns a list of certificates that includes both source and target, it could do something like this:
```
    private Future<Integer> generateAuthHash(String namespace, KafkaMirrorMaker2Cluster mirrorMaker2Cluster, List<String> certs) {
        Promise<Integer> authHash = Promise.promise();
        int certsHash = certs.stream().mapToInt(String::hashCode).sum();

        Future.join(mirrorMaker2Cluster
                        .clusters()
                        .stream()
                        .map(cluster -> ReconcilerUtils.authTlsHash(secretOperations, namespace, cluster.getAuthentication(), emptyList())).collect(Collectors.toList())
                )
                .onSuccess(authHashes -> {
                    int hash = authHashes.<Integer>list()
                        .stream()
                        .mapToInt(i -> i)
                        .sum();
                    authHash.complete(hash + certsHash);
                }).onFailure(authHash::fail);

        return authHash.future();
    }
```
tlsTrustedCertsSecret could create a separate list to return using the secret data from each source and target cluster.
What do you think?

venkatesh2090 commented:
Yes, this makes sense 👍️. Thanks for the clarification. Will chew at it over the new few days.

venkatesh2090 committed:
refactor hash function. ordering of certs shouldn't affect auth hash value.

tlsTrustedCertsSecret returns a Collection which is used to calculate the auth hash

Signed-off-by: Venkatesh Kannan <venkatesprasad512@gmail.com>

venkatesh2090 committed:
refactor tlsTrustedCertsSecret - don't need multiple signatures. Only one is used outside the class.

Signed-off-by: Venkatesh Kannan <venkatesprasad512@gmail.com>

venkatesh2090 committed:
refactor changes after rebasing

Signed-off-by: Venkatesh Kannan <venkatesprasad512@gmail.com>

tinaselenge commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaMirrorMaker2AssemblyOperator.java:
```diff
@@ -242,7 +243,7 @@ protected Future<Void> tlsTrustedCertsSecret(Reconciliation reconciliation, Stri
                                             KafkaConnectResources.internalTlsTrustedCertsSecretName(mirrorMaker2Cluster.getCluster())
                                     )))
-                            .mapEmpty();
+                            .map(secretData.values());
```
Does this mean we are returning target certificates twice now and so auth hash is being calculated twice for the target cluster?  because the map contains connect key mapped to target certificates, source alias mapped to source certificates, target alias mapped to target certificates right? Maybe a creating a separate list  from the Map would be better so we can return target certificates only once. What do you think?

venkatesh2090 replied:
> Does this mean we are returning target certificates twice now and so auth hash is being calculated twice for the target cluster?

Yes, I think that's correct.

What if we just mutate the `secretData` to remove the connect key?

```suggestion
                            .map(s -> secretData.remove(KafkaConnectCluster.KAFKA_CONNECT_CERTIFICATES_KEY))
                            .map(secretData.values());
```
or
```suggestion
                            .map(s -> {
                                secretData.remove(KafkaConnectCluster.KAFKA_CONNECT_CERTIFICATES_KEY);
                                return secretData.values();
                            });
```

I don't want to create a new object just so we can have 1 less item. Might as well remove it from `secretData`.

If I am understanding async operations correctly, `secretData` should have been reconciled with the secret by the time `map()` is called on the `Future`?

tinaselenge replied:
yes I think that would work. The other alternative is to use the Set to deduplicate the certificate values. e.g:
```
        return Future.join(new ArrayList<>(certificatesFutures.values()))
                .compose(i -> {
                    Map<String, String> secretData = new HashMap<>();
                    Set<String> certificates = new HashSet<>();
                    for (Map.Entry<String, Future<String>> entry : certificatesFutures.entrySet()) {
                        String certificate = entry.getValue().result();
                        if (certificate != null && !certificate.isEmpty()) {
                            secretData.put(entry.getKey(), Util.encodeToBase64(certificate));
                            certificates.add(certificate);
                        }
                    }

                    return VertxUtil.toFuture(secretOperations.reconcile(
                                    reconciliation,
                                    namespace,
                                    KafkaConnectResources.internalTlsTrustedCertsSecretName(mirrorMaker2Cluster.getCluster()),
                                    mirrorMaker2Cluster.generateTlsTrustedCertsSecret(
                                            secretData,
                                            KafkaConnectResources.internalTlsTrustedCertsSecretName(mirrorMaker2Cluster.getCluster())
                                    )))
                            .map(ignore -> certificates);
                });
    }
```
This might be slightly clearer on the intent but up to you.

venkatesh2090 replied:
Ok. I think I'll just use the Set like you suggested. It is indeed hard to justify remove an element in the comments. Not that big of a deal.

tinaselenge replied:
Thanks! Just curious, why you have decided to bundle them before adding it to the set though? For bridge and connect, you return all the certificates as separate elements in a Collection. For this, couldn't we just do `certs.addAll(certificates)`? Also can we add a comment somewhere to explain that we are doing this to calculate the hash of each certificate later on?

venkatesh2090 replied:
Hmm, I think I was just doing it per cluster because that's how it is serialised for the Secret. Even in the operators I am bundling it

https://github.com/venkatesh2090/strimzi-kafka-operator/blob/3bad5e223a919bcfa4004c974a20a339d8289115/cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/AbstractConnectOperator.java#L294-L295

Because before this change the bundling was done by ReconcilerUtils.trustedCertificates.

https://github.com/strimzi/strimzi-kafka-operator/blob/24fe3c0463e671be1d3131468e81b0a0d1262020/cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/ReconcilerUtils.java#L514

tinaselenge replied:
Yes they are bundled for secrets which is fine. For auth hash, I thought we wanted to calculate hash of the certificates individually and then sum them. You are doing this correctly for Bridge and Connect by returning collection of certificates rather than bundling them, I am wondering why this is different MM2.

> Because before this change the bundling was done by ReconcilerUtils.trustedCertificates.

Yes that's we do for reconciling the internal secrets which is fine. But today ReconcilerUtils.authTlsHash() method iterates through each certificate source, fetches its certificate and calculates the hash before summing them for all certificate sources. Whereas here, you are bundling certificates from all the certificate sources before calculating hash on the bundle, not on the individual certificates anymore.

venkatesh2090 replied:
Ooh. Makes sense now. Yes, will update

tinaselenge commented on cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/ReconcilerUtilsTest.java:
```diff
@@ -461,75 +433,13 @@ void getHashFailure() {
         when(secretOps.getAsync(eq(namespace), eq("top-secret-pwd"))).thenReturn(CompletableFuture.completedFuture(null));
 
-        Future<Integer> res = ReconcilerUtils.authTlsHash(secretOps, "ns", kcu, List.of());
+        Future<Integer> res = ReconcilerUtils.authTlsHash(secretOps, "ns", kcu, null);
```
what is the reason for this change?

venkatesh2090 replied:
Not sure :thinking:. I'll revert these changes back. I can't remember, this was a while ago. But looking at the commit history, it seems I changed it when I `authTlsHash` signature to accept a `String` cert bundle.

https://github.com/strimzi/strimzi-kafka-operator/pull/12241/changes/128da128c7bb48ccaf2e6fe4665d7b1ec635adae#diff-99728079ba041e1b6528b53dcd7d477fad3af5e8045e4a9bad5f30a9ff774551R460

In that commit I forgot to add a `String.isBlank` check. Hence the tests pass in `null` instead of a blank `String`. I later added that as well

https://github.com/strimzi/strimzi-kafka-operator/pull/12241/changes/a3c14f2f6a77263482819eafaefa83592a00ff2c#diff-99728079ba041e1b6528b53dcd7d477fad3af5e8045e4a9bad5f30a9ff774551R462

But forgot to change that null to a `""`.

And after the current change, since I have passed in `null`, it never failed to compile and I missed this completely.

tinaselenge commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/AbstractConnectOperator.java:
```diff
@@ -275,9 +276,12 @@ protected Future<ReconcileResult<NetworkPolicy>> connectNetworkPolicy(Reconcilia
+     * @return Future which completes when the reconciliation is done.
      */
-    protected Future<Void> tlsTrustedCertsSecret(Reconciliation reconciliation, String namespace, KafkaConnectCluster connect) {
+    protected Future<? extends Collection<String>> tlsTrustedCertsSecret(Reconciliation reconciliation, String namespace, KafkaConnectCluster connect) {
```
any reason why Collection<String> is being used instead of List and so also why wildcard for the Future being used? Are we expecting to receive different type of collection?

venkatesh2090 replied:
Using this because Map.values() used in https://github.com/strimzi/strimzi-kafka-operator/pull/12241/changes#diff-c6d9b371d64b172f4e174d61f4a6b81ed129010c41a016f0bbf29dddc62df009R246 returns a `Collection` (an I didn't want to create a copy just to convert it back to a List) but other implementations of that method returned a `List`. Can't remember why I added the wildcard though. I did have some trouble with `ReconcilerUtils.trustedCertificates`. But seems like it is gone after this commit?

https://github.com/strimzi/strimzi-kafka-operator/pull/12241/changes/6288302f973a90f87247dd5c58ad38d3f32a25e5

Need to validate. I'll revert it back to a `Collection` if it works without the wildcard.

venkatesh2090 committed:
change tlsTrustedCertsSecret signature
undo ReconcilerUtilsTest changes

Signed-off-by: Venkatesh Kannan <venkatesprasad512@gmail.com>

venkatesh2090 committed:
modify tlsTrustedCertsSecret to only return target cert once. This impacts the hash function because before this target cert was being used twice in it.

Signed-off-by: Venkatesh Kannan <venkatesprasad512@gmail.com>

katheris commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/AbstractConnectOperator.java:
```diff
@@ -275,9 +276,12 @@ protected Future<ReconcileResult<NetworkPolicy>> connectNetworkPolicy(Reconcilia
-     * @return  Future which completes when the reconciliation is done
+     * @param reconciliation The reconcilation.
+     * @param namespace Namespace of the Connect cluster.
+     * @param connect KafkaConectCluster object.
```
```suggestion
     * @param connect KafkaConnectCluster object.
```

katheris replied:
```suggestion
     * @param connect KafkaConnectCluster object.
```

katheris commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/ReconcilerUtils.java:
```suggestion
     * Gets trusted certificates from Secrets and merges them into a single List of Strings.
```

katheris commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/ReconcilerUtils.java:
```diff
@@ -510,7 +508,7 @@ public static Future<String> trustedCertificates(Reconciliation reconciliation,
                         } else {
-                            return Future.succeededFuture(String.join("\n", certificates.list()));
+                            return Future.succeededFuture(certificates.list());
                         }
```
Do we still need this compose? Why not return the list directly since returning an empty list is probably preferable to returning null in the case there are no certificates.

venkatesh2090 replied:
Oh didn't realise. Yes can do that

venkatesh2090 replied:
Actually I should probably switch it to a null check instead. Because this is what `list()` returns

> a list of the current completed values. If one future is not yet resolved or is failed, null will be used

It is nullable. What do you think?

venkatesh2090 replied:
Just read the documentation again... I think they mean the value of the future that failed will be null, not the return value of list(). Will update as you suggested

katheris commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectAssemblyOperator.java:
```diff
@@ -288,13 +288,12 @@ private Future<Void> updateConnectorsThatConnectClusterWasDeleted(Reconciliation
-        KafkaClientAuthentication auth = kafkaConnectSpec.getAuthentication();
-        List<CertSecretSource> trustedCertificates = kafkaConnectSpec.getTls() == null ? Collections.emptyList() : kafkaConnectSpec.getTls().getTrustedCertificates();
-        return ReconcilerUtils.authTlsHash(secretOperations, namespace, auth, trustedCertificates);
+    private Future<Integer> generateAuthHash(String namespace, KafkaClientAuthentication authentication, Collection<String> certificates) {
```
Since this now calls straight through to ReconcilerUtils we can do this inline rather than having a separate method.

venkatesh2090 replied:
Can do. Makes sense

katheris reviewed (commented):
Hey @venkatesh2090 I've had another go through and added a couple of extra comments, but I'm unclear why the type is not List<String> throughout and why the wildcard is needed for the Map of futures, did you confirm why you needed to use Collection and the wildcard?

venkatesh2090 commented:
> Hey @venkatesh2090 I've had another go through and added a couple of extra comments, but I'm unclear why the type is not List throughout and why the wildcard is needed for the Map of futures, did you confirm why you needed to use Collection and the wildcard?

The wildcard for Map I must have missed. I'll revert it. As for the collection - this is where I explained https://github.com/strimzi/strimzi-kafka-operator/pull/12241#discussion_r3521720064

venkatesh2090 committed:
Address some comments:

- modify comments
- remove and refactor methods
- refactor tls cert secrets Map type

Signed-off-by: Venkatesh Kannan <venkatesprasad512@gmail.com>

venkatesh2090 committed:
refactored KafkaMirrorMaker2AssemblyOperator.tlsTrustedCertsSecret returned Set
refactored ReconcilerUtils.trustedCertificates cert composing Futures

Signed-off-by: Venkatesh Kannan <venkatesprasad512@gmail.com>

katheris commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaMirrorMaker2AssemblyOperator.java:
```diff
@@ -242,7 +247,9 @@ protected Future<Void> tlsTrustedCertsSecret(Reconciliation reconciliation, Stri
                                             KafkaConnectResources.internalTlsTrustedCertsSecretName(mirrorMaker2Cluster.getCluster())
                                     )))
-                            .mapEmpty();
+                            // This will be used to calculate auth hash. Hence, a unique collection
```
Can you move this comment next to where we instantiate the certs field? Otherwise I suspect it's likely to get lost in refactors.

katheris reviewed (commented):
@venkatesh2090 I am happy with the logic now, just two small comments left for me, moving a comment and adding those tests back in.

venkatesh2090 committed:
Move comment to a sensible location

Signed-off-by: Venkatesh Kannan <venkatesprasad512@gmail.com>

venkatesh2090 committed:
Replacing getHashPatternNotMatching for the equivalent trustedCertificates test

Signed-off-by: Venkatesh Kannan <venkatesprasad512@gmail.com>

tinaselenge commented:
/gha run pipeline=regression

katheris commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaMirrorMaker2AssemblyOperator.java:
It's annoying I didn't spot this before but we are missing some javadoc here for the new param `certs`

venkatesh2090 replied:
Updated

katheris commented on cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/ReconcilerUtilsTest.java:
```diff
@@ -788,6 +699,28 @@ void testTrustedCertificatesInvalidCertificate(VertxTestContext context) {
         }));
     }
 
+    @Test
```
I think we're still missing one test case, the replacement for `getHashForPattern`. Perhaps the `testTrustedCertificates` test can just be updated so that `cert-secret` contains two `.crt` files, then the `assertThat` line above would be:
```
assertThat(res, hasItems(DUMMY_CERT + "\n" + DUMMY_CERT, DUMMY_CERT, DUMMY_CERT));
```

venkatesh2090 replied:
Updated this as well. I didn't pay attention to the number of certs in the Secret. Thanks

katheris reviewed (commented):
@venkatesh2090 Thanks for updating the PR, I'm happy with the updated test, but I think you're still missing one and I also spotted a missing Javadoc

venkatesh2090 committed:
Add more tests and fix javadoc

Signed-off-by: Venkatesh Kannan <venkatesprasad512@gmail.com>

katheris reviewed (approved):
Thanks @venkatesh2090 for all your work on this, I'm happy with these changes now

tinaselenge reviewed (approved):
I had a look again, it looks good to me now. Thank you for your contribution.
